src/index.js

```diff
@@ -0,0 +1,226 @@
+'use strict'
+
+const EventEmitter = require('events')
+const pull = require('pull-stream/pull')
+const empty = require('pull-stream/sources/empty')
+const asyncEach = require('async/each')
+const debug = require('debug')
+
+const Peer = require('./peer')
+const message = require('./message')
+
+const nextTick = require('async/nextTick')
+
+/**
+ * PubsubBaseProtocol handles the peers and connections logic for pubsub routers
+ */
+class PubsubBaseProtocol extends EventEmitter {
+  /**
+   * @param {String} debugName
+   * @param {String} multicodec
+   * @param {Object} libp2p
+   * @constructor
+   */
+  constructor (debugName, multicodec, libp2p) {
+    super()
+
+    this.log = debug(debugName)
+    this.log.err = debug(`${debugName}:error`)
+    this.multicodec = multicodec
+    this.libp2p = libp2p
+    this.started = false
+
+    /**
+     * Map of peers.
+     *
+     * @type {Map<string, Peer>}
+     */
+    this.peers = new Map()
+
+    // Dials that are currently in progress
+    this._dials = new Set()
+
+    this._onConnection = this._onConnection.bind(this)
+    this._dialPeer = this._dialPeer.bind(this)
+  }
+
+  _addPeer (peer) {
+    const id = peer.info.id.toB58String()
+
+    /*
+      Always use an existing peer.
+
+      What is happening here is: "If the other peer has already dialed to me, we already have
+      an establish link between the two, what might be missing is a
+      Connection specifically between me and that Peer"
+     */
+    let existing = this.peers.get(id)
+    if (!existing) {
+      this.log('new peer', id)
+      this.peers.set(id, peer)
+      existing = peer
+
+      peer.once('close', () => this._removePeer(peer))
+    }
+    ++existing._references
+
+    return existing
+  }
+
+  _removePeer (peer) {
+    const id = peer.info.id.toB58String()
+
+    this.log('remove', id, peer._references)
+    // Only delete when no one else is referencing this peer.
+    if (--peer._references === 0) {
+      this.log('delete peer', id)
+      this.peers.delete(id)
+    }
+
+    return peer
+  }
+
+  _dialPeer (peerInfo, callback) {
+    callback = callback || function noop () { }
+    const idB58Str = peerInfo.id.toB58String()
+
+    // If already have a PubSub conn, ignore
+    const peer = this.peers.get(idB58Str)
+    if (peer && peer.isConnected) {
+      return nextTick(() => callback())
+    }
+
+    // If already dialing this peer, ignore
+    if (this._dials.has(idB58Str)) {
+      this.log('already dialing %s, ignoring dial attempt', idB58Str)
+      return nextTick(() => callback())
+    }
+    this._dials.add(idB58Str)
+
+    this.log('dialing %s', idB58Str)
+    this.libp2p.dialProtocol(peerInfo, this.multicodec, (err, conn) => {
+      this.log('dial to %s complete', idB58Str)
+
+      // If the dial is not in the set, it means that floodsub has been
+      // stopped
+      const floodsubStopped = !this._dials.has(idB58Str)
```

Comment:
`floodsubStopped`; change floodsub references to generic pubsub? Comments should be updated too.